File: backend/tests/rss.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import { fetchAllNewReviews } from "../src/infra/rss";
import type { AppConfig } from "../src/domain/types";

const APP: AppConfig = { id: "595068606", name: "Headspace", country: "us" };

function makeEntry(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    id: { label: "default-id" },
    author: { name: { label: "Bob" } },
    "im:rating": { label: "5" },
    title: { label: "Great" },
    content: { label: "Love it" },
    updated: { label: "2026-05-19T10:00:00-07:00" },
    ...overrides,
  };
}

type MockResponse =
  | { ok?: boolean; status?: number; body: unknown }
  | { throws: Error }
  | { jsonThrows: true };

function mockFetch(...responses: MockResponse[]): void {
  const spy = vi.spyOn(global, "fetch");
  for (const r of responses) {
    if ("throws" in r) {
      spy.mockRejectedValueOnce(r.throws);
    } else if ("jsonThrows" in r) {
      spy.mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => {
          throw new Error("invalid JSON");
        },
      } as unknown as Response);
    } else {
      spy.mockResolvedValueOnce({
        ok: r.ok ?? true,
        status: r.status ?? 200,
        json: async () => r.body,
      } as unknown as Response);
    }
  }
}

describe("fetchAllNewReviews", () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe("parsing", () => {
    it("maps Apple's labeled shape into Review objects", async () => {
      mockFetch(
        {
          body: {
            feed: {
              entry: [makeEntry({ id: { label: "r1" }, "im:rating": { label: "4" } })],
            },
          },
        },
        { body: { feed: { entry: [] } } },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 2);

      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data).toHaveLength(1);
      expect(result.data[0]).toMatchObject({
        id: "r1",
        appId: "595068606",
        author: "Bob",
        score: 4,
        title: "Great",
        content: "Love it",
        submittedAt: "2026-05-19T10:00:00-07:00",
      });
    });

    it("handles `entry` as a single object (Apple's one-review quirk)", async () => {
      mockFetch({ body: { feed: { entry: makeEntry() } } }, { body: { feed: { entry: [] } } });

      const result = await fetchAllNewReviews(APP, new Set(), 2);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data).toHaveLength(1);
    });

    it("returns empty array when feed has no entries", async () => {
      mockFetch({ body: { feed: {} } });

      const result = await fetchAllNewReviews(APP, new Set(), 1);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data).toEqual([]);
    });
  });

  describe("filtering bad rows", () => {
    it("skips entries without im:rating (app-metadata rows)", async () => {
      mockFetch(
        {
          body: {
            feed: {
              entry: [
                makeEntry({ id: { label: "good" } }),
                { ...makeEntry({ id: { label: "no-rating" } }), "im:rating": undefined },
              ],
            },
          },
        },
        { body: { feed: { entry: [] } } },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 2);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data.map((r) => r.id)).toEqual(["good"]);
    });

    it("skips non-integer ratings", async () => {
      mockFetch(
        {
          body: {
            feed: {
              entry: [
                makeEntry({ id: { label: "five" }, "im:rating": { label: "5" } }),
                makeEntry({ id: { label: "decimal" }, "im:rating": { label: "4.5" } }),
                makeEntry({ id: { label: "garbage" }, "im:rating": { label: "abc" } }),
                makeEntry({ id: { label: "empty" }, "im:rating": { label: "" } }),
              ],
            },
          },
        },
        { body: { feed: { entry: [] } } },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 2);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data.map((r) => r.id)).toEqual(["five"]);
    });

    it("skips ratings outside the 1-5 range", async () => {
      mockFetch(
        {
          body: {
            feed: {
              entry: [
                makeEntry({ id: { label: "ok" }, "im:rating": { label: "3" } }),
                makeEntry({ id: { label: "zero" }, "im:rating": { label: "0" } }),
                makeEntry({ id: { label: "six" }, "im:rating": { label: "6" } }),
                makeEntry({ id: { label: "negative" }, "im:rating": { label: "-1" } }),
              ],
            },
          },
        },
        { body: { feed: { entry: [] } } },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 2);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data.map((r) => r.id)).toEqual(["ok"]);
    });
  });

  describe("error handling", () => {
    it("returns fail with upstream status on non-2xx HTTP response", async () => {
      mockFetch({ ok: false, status: 503, body: {} });

      const result = await fetchAllNewReviews(APP, new Set(), 1);
      if (result.success) throw new Error("expected failure");
      expect(result.status).toBe(503);
      expect(result.error).toContain("503");
    });

    it("returns fail with 504 on network error", async () => {
      mockFetch({ throws: new Error("ECONNREFUSED") });

      const result = await fetchAllNewReviews(APP, new Set(), 1);
      if (result.success) throw new Error("expected failure");
      expect(result.status).toBe(504);
      expect(result.error).toContain("Network error");
    });

    it("returns fail with 502 when response isn't valid JSON", async () => {
      mockFetch({ jsonThrows: true });

      const result = await fetchAllNewReviews(APP, new Set(), 1);
      if (result.success) throw new Error("expected failure");
      expect(result.status).toBe(502);
      expect(result.error).toContain("Invalid JSON");
    });
  });

  describe("pagination", () => {
    it("walks multiple pages and accumulates results", async () => {
      mockFetch(
        { body: { feed: { entry: [makeEntry({ id: { label: "p1-1" } })] } } },
        { body: { feed: { entry: [makeEntry({ id: { label: "p2-1" } })] } } },
        { body: { feed: { entry: [] } } },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 5);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data.map((r) => r.id)).toEqual(["p1-1", "p2-1"]);
    });

    it("stops at the first known ID within a page", async () => {
      const known = new Set(["known"]);
      mockFetch({
        body: {
          feed: {
            entry: [
              makeEntry({ id: { label: "new-1" } }),
              makeEntry({ id: { label: "known" } }),
              makeEntry({ id: { label: "would-be-included-if-not-known" } }),
            ],
          },
        },
      });

      const result = await fetchAllNewReviews(APP, known, 5);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data.map((r) => r.id)).toEqual(["new-1"]);
    });

    it("respects the maxPages cap", async () => {
      mockFetch(
        { body: { feed: { entry: [makeEntry({ id: { label: "p1" } })] } } },
        { body: { feed: { entry: [makeEntry({ id: { label: "p2" } })] } } },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 2);
      if (!result.success) throw new Error(`expected success, got: ${result.error}`);
      expect(result.data.map((r) => r.id)).toEqual(["p1", "p2"]);
    });

    it("bubbles failure from a mid-page fetch", async () => {
      mockFetch(
        { body: { feed: { entry: [makeEntry({ id: { label: "p1" } })] } } },
        { ok: false, status: 429, body: {} },
      );

      const result = await fetchAllNewReviews(APP, new Set(), 5);
      if (result.success) throw new Error("expected failure");
      expect(result.status).toBe(429);
    });
  });
});

File: backend/internal/rss/rss_test.go
package rss

import (
	"encoding/json"
	"testing"
)

func validEntry() rawEntry {
	return rawEntry{
		ID:       labeled{Label: "r1"},
		Author:   rawAuthor{Name: labeled{Label: "Bob"}},
		ImRating: &labeled{Label: "5"},
		Title:    labeled{Label: "Great"},
		Content:  labeled{Label: "Love it"},
		Updated:  labeled{Label: "2026-05-19T10:00:00-07:00"},
	}
}

func TestToReview_ValidEntry(t *testing.T) {
	r := toReview(validEntry(), "app1")
	if r == nil {
		t.Fatal("expected a review, got nil")
	}
	if r.ID != "r1" || r.AppID != "app1" || r.Author != "Bob" || r.Score != 5 {
		t.Errorf("unexpected review: %+v", r)
	}
	if r.SubmittedAt != "2026-05-19T10:00:00-07:00" {
		t.Errorf("unexpected submittedAt: %q", r.SubmittedAt)
	}
}

func TestToReview_SkipsWithoutRating(t *testing.T) {
	e := validEntry()
	e.ImRating = nil
	if r := toReview(e, "app1"); r != nil {
		t.Errorf("expected nil for missing im:rating, got %+v", r)
	}
}

func TestToReview_SkipsEmptyRating(t *testing.T) {
	e := validEntry()
	e.ImRating = &labeled{Label: ""}
	if r := toReview(e, "app1"); r != nil {
		t.Errorf("expected nil for empty rating, got %+v", r)
	}
}

func TestToReview_SkipsNonIntegerRating(t *testing.T) {
	cases := []string{"abc", "4.5", "", " "}
	for _, c := range cases {
		e := validEntry()
		e.ImRating = &labeled{Label: c}
		if r := toReview(e, "app1"); r != nil {
			t.Errorf("rating %q: expected nil, got %+v", c, r)
		}
	}
}

func TestToReview_SkipsOutOfRangeRating(t *testing.T) {
	for _, raw := range []string{"0", "6", "-1", "10"} {
		e := validEntry()
		e.ImRating = &labeled{Label: raw}
		if r := toReview(e, "app1"); r != nil {
			t.Errorf("rating %q: expected nil, got %+v", raw, r)
		}
	}
}

func TestToReview_AcceptsAllValidRatings(t *testing.T) {
	for _, raw := range []string{"1", "2", "3", "4", "5"} {
		e := validEntry()
		e.ImRating = &labeled{Label: raw}
		if r := toReview(e, "app1"); r == nil {
			t.Errorf("rating %q: expected valid review, got nil", raw)
		}
	}
}

func TestToReview_SkipsMissingRequiredFields(t *testing.T) {
	cases := map[string]func(e *rawEntry){
		"missing id":      func(e *rawEntry) { e.ID.Label = "" },
		"missing author":  func(e *rawEntry) { e.Author.Name.Label = "" },
		"missing content": func(e *rawEntry) { e.Content.Label = "" },
		"missing updated": func(e *rawEntry) { e.Updated.Label = "" },
	}
	for name, mutate := range cases {
		e := validEntry()
		mutate(&e)
		if r := toReview(e, "app1"); r != nil {
			t.Errorf("%s: expected nil, got %+v", name, r)
		}
	}
}

func TestNormalizeEntries_Empty(t *testing.T) {
	got, err := normalizeEntries(nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != nil {
		t.Errorf("expected nil, got %+v", got)
	}
}

func TestNormalizeEntries_Array(t *testing.T) {
	raw := json.RawMessage(`[
		{"id":{"label":"a"}, "author":{"name":{"label":"x"}}, "im:rating":{"label":"5"}, "title":{"label":"t"}, "content":{"label":"c"}, "updated":{"label":"u"}},
		{"id":{"label":"b"}, "author":{"name":{"label":"y"}}, "im:rating":{"label":"4"}, "title":{"label":"t"}, "content":{"label":"c"}, "updated":{"label":"u"}}
	]`)
	got, err := normalizeEntries(raw)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(got) != 2 {
		t.Fatalf("expected 2 entries, got %d", len(got))
	}
	if got[0].ID.Label != "a" || got[1].ID.Label != "b" {
		t.Errorf("unexpected IDs: %s, %s", got[0].ID.Label, got[1].ID.Label)
	}
}

func TestNormalizeEntries_SingleObject(t *testing.T) {
	raw := json.RawMessage(`{"id":{"label":"only"}, "author":{"name":{"label":"x"}}, "im:rating":{"label":"5"}, "title":{"label":"t"}, "content":{"label":"c"}, "updated":{"label":"u"}}`)
	got, err := normalizeEntries(raw)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(got) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(got))
	}
	if got[0].ID.Label != "only" {
		t.Errorf("unexpected ID: %s", got[0].ID.Label)
	}
}

func TestNormalizeEntries_Invalid(t *testing.T) {
	raw := json.RawMessage(`not json`)
	if _, err := normalizeEntries(raw); err == nil {
		t.Error("expected an error for invalid JSON, got nil")
	}
}

func TestBuildFeedURL(t *testing.T) {
	got := buildFeedURL("595068606", "us", 1)
	want := "https://itunes.apple.com/us/rss/customerreviews/id=595068606/sortBy=mostRecent/page=1/json"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}
